File: go/test/get_location_by_postal_code_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/zippopotamus-zip-code-sdk/go"
	"github.com/voxgig-sdk/zippopotamus-zip-code-sdk/go/core"

	vs "github.com/voxgig-sdk/zippopotamus-zip-code-sdk/go/utility/struct"
)

func TestGetLocationByPostalCodeEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.GetLocationByPostalCode(nil)
		if ent == nil {
			t.Fatal("expected non-nil GetLocationByPostalCodeEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := get_location_by_postal_codeBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "get_location_by_postal_code." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set ZIPPOPOTAMUSZIPCODE_TEST_GET_LOCATION_BY_POSTAL_CODE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		getLocationByPostalCodeRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.get_location_by_postal_code", setup.data)))
		var getLocationByPostalCodeRef01Data map[string]any
		if len(getLocationByPostalCodeRef01DataRaw) > 0 {
			getLocationByPostalCodeRef01Data = core.ToMapAny(getLocationByPostalCodeRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = getLocationByPostalCodeRef01Data

		// LIST
		getLocationByPostalCodeRef01Ent := client.GetLocationByPostalCode(nil)
		getLocationByPostalCodeRef01Match := map[string]any{
			"country": setup.idmap["country01"],
			"postal_code": setup.idmap["postal_code01"],
		}

		getLocationByPostalCodeRef01ListResult, err := getLocationByPostalCodeRef01Ent.List(getLocationByPostalCodeRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, getLocationByPostalCodeRef01ListOk := getLocationByPostalCodeRef01ListResult.([]any)
		if !getLocationByPostalCodeRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", getLocationByPostalCodeRef01ListResult)
		}

	})
}

func get_location_by_postal_codeBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "get_location_by_postal_code", "GetLocationByPostalCodeTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read get_location_by_postal_code test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse get_location_by_postal_code test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"get_location_by_postal_code01", "get_location_by_postal_code02", "get_location_by_postal_code03", "country01", "postal_code01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("ZIPPOPOTAMUSZIPCODE_TEST_GET_LOCATION_BY_POSTAL_CODE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"ZIPPOPOTAMUSZIPCODE_TEST_GET_LOCATION_BY_POSTAL_CODE_ENTID": idmap,
		"ZIPPOPOTAMUSZIPCODE_TEST_LIVE":      "FALSE",
		"ZIPPOPOTAMUSZIPCODE_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["ZIPPOPOTAMUSZIPCODE_TEST_GET_LOCATION_BY_POSTAL_CODE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["ZIPPOPOTAMUSZIPCODE_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewZippopotamusZipCodeSDK(core.ToMapAny(mergedOpts))
	}

	live := env["ZIPPOPOTAMUSZIPCODE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["ZIPPOPOTAMUSZIPCODE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/get_location_by_postal_code_entity_test.lua
-- GetLocationByPostalCode entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("zippopotamus-zip-code_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("GetLocationByPostalCodeEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:GetLocationByPostalCode(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = get_location_by_postal_code_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"list"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "get_location_by_postal_code." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set ZIPPOPOTAMUSZIPCODE_TEST_GET_LOCATION_BY_POSTAL_CODE_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local get_location_by_postal_code_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.get_location_by_postal_code")))
    local get_location_by_postal_code_ref01_data = nil
    if #get_location_by_postal_code_ref01_data_raw > 0 then
      get_location_by_postal_code_ref01_data = helpers.to_map(get_location_by_postal_code_ref01_data_raw[1][2])
    end

    -- LIST
    local get_location_by_postal_code_ref01_ent = client:GetLocationByPostalCode(nil)
    local get_location_by_postal_code_ref01_match = {
      ["country"] = setup.idmap["country01"],
      ["postal_code"] = setup.idmap["postal_code01"],
    }

    local get_location_by_postal_code_ref01_list_result, err = get_location_by_postal_code_ref01_ent:list(get_location_by_postal_code_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(get_location_by_postal_code_ref01_list_result)

  end)
end)

function get_location_by_postal_code_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/get_location_by_postal_code/GetLocationByPostalCodeTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read get_location_by_postal_code test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "get_location_by_postal_code01", "get_location_by_postal_code02", "get_location_by_postal_code03", "country01", "postal_code01" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("ZIPPOPOTAMUSZIPCODE_TEST_GET_LOCATION_BY_POSTAL_CODE_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["ZIPPOPOTAMUSZIPCODE_TEST_GET_LOCATION_BY_POSTAL_CODE_ENTID"] = idmap,
    ["ZIPPOPOTAMUSZIPCODE_TEST_LIVE"] = "FALSE",
    ["ZIPPOPOTAMUSZIPCODE_TEST_EXPLAIN"] = "FALSE",
    ["ZIPPOPOTAMUSZIPCODE_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["ZIPPOPOTAMUSZIPCODE_TEST_GET_LOCATION_BY_POSTAL_CODE_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["ZIPPOPOTAMUSZIPCODE_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["ZIPPOPOTAMUSZIPCODE_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["ZIPPOPOTAMUSZIPCODE_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["ZIPPOPOTAMUSZIPCODE_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
